kernel/posix/pthread_common.c

```diff
@@ -18,3 +19,34 @@ void ready_one_thread(_wait_q_t *wq)
 		_ready_thread(th);
 	}
 }
+
+s64_t timespec_to_timeoutms(const struct timespec *abstime)
+{
+	s64_t milli_secs;
+	s32_t secs;
+	struct timespec curtime;
+
+	/* FIXME: Zephyr does have CLOCK_REALTIME to get time.
+	 * As per POSIX standard time should be calculated wrt CLOCK_REALTIME.
+	 * Zephyr deviates from POSIX 1003.1 standard on this aspect.
+	 */
+	clock_gettime(CLOCK_MONOTONIC, &curtime);
+	secs = abstime->tv_sec - curtime.tv_sec;
+
+	if (abstime->tv_sec < curtime.tv_sec ||
+	    (secs == 0 && abstime->tv_nsec <= curtime.tv_nsec)) {
+		milli_secs = K_NO_WAIT;
+	} else {
+		milli_secs = (abstime->tv_nsec - curtime.tv_nsec) /
+			      NSEC_PER_MSEC;
+
+		if (milli_secs < 0) {
+			milli_secs += MSEC_PER_SEC;
+			secs -= 1;
+		}
+
+		milli_secs += (long)secs * MSEC_PER_SEC;
+	}
+
+	return milli_secs;
```

Comment:
I probably left too many comments in this function.  Broadly, it's just too complicated. You don't want to be branching at all before the final line.  Compute a single, signed timeout delta and return it if it's positive otherwise K_NO_WAIT.

Reply:
Sure I will try to simplify this function. This function i coded long back and i thought of using same.